File: test/testMatrix.cpp
/*******************************************************************************
 * This file is part of CosTuuM
 * Copyright (C) 2019 Bert Vandenbroucke (bert.vandenbroucke@gmail.com)
 *
 * CosTuuM is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * CosTuuM is distributed in the hope that it will be useful, but WITOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with CosTuuM. If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

/**
 * @file testMatrix.cpp
 *
 * @brief Unit test for the Matrix class and its member functions.
 *
 * @author Bert Vandenbroucke (bert.vandenbroucke@ugent.be)
 */

#include "Configuration.hpp"
#include "Matrix.hpp"
#include <iostream>

#if defined(HAVE_MULTIPRECISION) && defined(HAVE_QUAD_PRECISION)
using namespace boost::multiprecision;
#else
using namespace std;
#endif

/**
 * @brief Unit test for the Matrix class and its member functions.
 *
 * @param argc Number of command line arguments (ignored).
 * @param argv Command line arguments (ignored).
 * @return Exit code: 0 on success.
 */
int main(int argc, char **argv) {

  Matrix<std::complex<float_type>> A(4, 4);
  A(0, 0) = 1.;
  A(0, 0).imag(1.);
  A(0, 1) = 2.;
  A(0, 1).imag(1.);
  A(0, 2) = 3.;
  A(0, 2).imag(1.);
  A(1, 0) = 4.;
  A(1, 0).imag(2.);
  A(1, 1) = 5.;
  A(1, 1).imag(1.);
  A(1, 2) = 6.;
  A(1, 2).imag(1.);
  A(2, 0) = 8.;
  A(2, 0).imag(3.);
  A(2, 1) = 9.;
  A(2, 1).imag(1.);
  A(2, 2).imag(-2.);

  // add some additional elements to the part of the matrix we don't care about
  A(3, 0) = 42.;
  A(3, 1) = 42.;
  A(3, 2) = 48.;
  A(3, 3) = 44.;
  A(0, 3) = 41.;
  A(1, 3) = 48.;
  A(2, 3) = 43.;

  for (uint_fast32_t i = 0; i < 3; ++i) {
    for (uint_fast32_t j = 0; j < 3; ++j) {
      std::cout << A(i, j) << " ";
    }
    std::cout << std::endl;
  }

  A.plu_inverse(3);

  A.binary_dump("test_matrix.dump");

  for (uint_fast32_t i = 0; i < 3; ++i) {
    for (uint_fast32_t j = 0; j < 3; ++j) {
      std::cout << A(i, j) << " ";
    }
    std::cout << std::endl;
  }

  return 0;
}

File: test/test_matrix.py
################################################################################
# This file is part of CosTuuM
# Copyright (C) 2019 Bert Vandenbroucke (bert.vandenbroucke@gmail.com)
#
# CosTuuM is free software: you can redistribute it and/or modify it under the
# terms of the GNU Affero General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# CosTuuM is distributed in the hope that it will be useful, but WITOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License
# along with CosTuuM. If not, see <http://www.gnu.org/licenses/>.
###############################################################################

##
# @file test_matrix.py
#
# @brief Script to read and analyse the output from the testMatrix unit test.
#
# We first reconstruct the output array from the binary dump file and then
# compare it with the same matrix inverted using numpy.linalg.inv.
#
# @author Bert Vandenbroucke (bert.vandenbroucke@ugent.be)
##

# load numpy for file input and matrix inversion
import numpy as np

# read the metadata from the dump file (matrix and element sizes)
metadata = np.fromfile("test_matrix.dump", dtype=np.uint32, count=3)
num_x = metadata[0]
num_y = metadata[1]

# read the actual matrix and reshape it to the correct shape
data = np.fromfile("test_matrix.dump", dtype=np.cdouble, offset=12)
data = data.reshape((num_x, num_y))

# construct the same initial matrix for the reference solution
A = np.array(
    [
        [1.0 + 1.0j, 2.0 + 1.0j, 3.0 + 1.0j],
        [4.0 + 2.0j, 5.0 + 1.0j, 6.0 + 1.0j],
        [8.0 + 3.0j, 9.0 + 1.0j, -2.0j],
    ],
    dtype=np.cdouble,
)
# invert it using numpy.linalg.inv
Ainv = np.linalg.inv(A)

# compare the unit test result and the numpy library result
for i in range(3):
    for j in range(3):
        reldiff = abs(data[i, j] - Ainv[i, j]) / abs(data[i, j] + Ainv[i, j])
        print("data:", data[i, j], "ref:", Ainv[i, j], "reldiff:", reldiff)
